src/main/java/com/github/packageurl/PackageURL.java

```diff
@@ -517,42 +533,39 @@ private static String toLowerCase(String s) {
         return new String(chars);
     }
 
-    /**
-     * Optionally decodes a String, if it's encoded. If String is not encoded,
-     * method will return the original input value.
-     *
-     * @param input the value String to decode
-     * @return a decoded String
-     */
-    private String percentDecode(final String input) {
-        if (input == null) {
-            return null;
-        }
-        final String decoded = uriDecode(input);
-        if (!decoded.equals(input)) {
-            return decoded;
+    public static String uriDecode(final String source) {
+        if (source == null || source.isEmpty()) {
+            return source;
         }
-        return input;
-    }
 
-    public static String uriDecode(String source) {
-        if (source == null) {
+        int percent = source.indexOf('%');
+
+        if (percent == -1) {
             return source;
         }
-        int length = source.length();
-        StringBuilder builder = new StringBuilder();
-        for (int i = 0; i < length; i++) {
-            if (source.charAt(i) == '%') {
-                String str = source.substring(i + 1, i + 3);
-                char c = (char) Integer.parseInt(str, 16);
-                builder.append(c);
-                i += 2;
-            }
-            else {
-                builder.append(source.charAt(i));
+
+        byte[] bytes = source.getBytes(StandardCharsets.UTF_8);
+        int length = bytes.length;
+        ByteArrayOutputStream buffer = new ByteArrayOutputStream(length);
+        buffer.write(bytes, 0, percent);
+
+        for (int i = percent; i < length; i++) {
+            int b = bytes[i];
+
+            if (b == '%') {
+                    if (i + 2 >= length) {
+                        return null;
+                    }
+
+                    int b1 = Character.digit(bytes[++i], 16);
+                    int b2 = Character.digit(bytes[++i], 16);
+                    buffer.write((char) ((b1 << 4) + b2));
+            } else {
+                buffer.write(b);
```

Comment:
I had a thought that we could improve the optimization of this even more:
```suggestion
        byte[] bytes = source.getBytes(StandardCharsets.UTF_8);
        int length = bytes.length;
        int writePos = percent;

        for (int i = percent; i < length; i++) {
            int b = bytes[i];

            if (b == '%') {
                if (i + 2 >= length) {
                    return null;
                }

                int b1 = Character.digit(bytes[++i], 16);
                int b2 = Character.digit(bytes[++i], 16);
                if (d1 == -1 || d2 == -1) {
                    return null;
                }
                bytes[writePos++] = (byte) ((d1 << 4) + d2);
            } else {
                bytes[writePos++] = b;
```

Reply:
My only other comment on this method would be to replace the `return null` with something like `throw new IllegalArgumentException("Invalid percent encoding at index " + i);`.

Reply:
@ppkarwasz I completely understand the benefits of proper shading. However, for a library like `packageurl-java` I am violently opposed to introducing any runtime dependencies - even if properly shaded.

Reply:
I agree that returning `null` was wrong. It's too bad that we can't throw a checked exception `MalformedPackageURLException` (or we can if you're OK with changing a `public` signature).

The other option is to be lenient and just ignore the `'%'`. I think I have seen some URI parsers do this.

Reply:
You can use `ValidationException` instead of `MalformedPackageURLException` as in other parts of the code.

Reply:
I think `ValidationException` is unchecked. I actually don't like this and think the code should be changed to throw a checked exception, but that's off-topic for this pull request.

But, good point in that it doesn't change the method signature. I am in favor of deprecating the `public` method since these methods should be `static` anyway. It probably wasn't meant to be public since only `decode` is `public` but not `encode`, too?

Reply:
So, I checked, and it's impossible to make it to this code with an invalid percent-encoding, as the `URI` constructor catches it. I added a test.

Reply:
You can reach the code through the deprecated, but still public `uriDecode` method.

Currently `new URI` detects invalid percent encoding sequences, but it might be useful to check it again, in case the code is optimized **not** to use `URI`.

Reply:
OK, the code now throws `ValidationException`. I added more tests.